### PR TITLE
Feature/issue#81 box

### PR DESCRIPTION
**Pull request by SC-AFI**
Releases must be distributed using [Box](https://www.box.com/). #81 
Example of the release: https://github.com/ReleaseDaemonTestOrganization/Test.Repo.123456/releases/tag/8.1.3.10

### CODE REVIEW

File: src/ReleaseDaemon/app.config [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (AlenPelin) [inline src/ReleaseDaemon/app.config]: Is this file really necessary?

File: src/ReleaseDaemon/Program.cs [hunk 2]
```diff
@@ -33,7 +33,8 @@ static void Main(string[] args)
         Certificate = ParseCertificate(args),
         NuGetSource = new NuGetSourceFactory(fileSystem).Parse(ParseSource(args)),
         RepositoryNamePrefixes = prefixes,
-        Credentials = ParseCredentials(args)
+        Credentials = ParseCredentials(args),
+        Storage = new PublicStorageFactory() {FileSystem = fileSystem }.Factory(GetPublicStorageConfiguration(args))
```
Comment (AlenPelin) [inline src/ReleaseDaemon/Program.cs]: Please introduce a variable as line is hard to read.

File: src/ReleaseDaemon/Model/PublicStorageInfoBase.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Sitecore.Support.ReleaseDaemon.Model
+{
+  using System.Collections.Generic;
+  using System.Threading.Tasks;
+
+  public abstract class PublicStorageInfoBase
```
Comment (AlenPelin) [inline src/ReleaseDaemon/Model/PublicStorageInfoBase.cs]: Why did abstract class was chosen instead of interface? There is not a big deal, I am just curious.

File: src/ReleaseDaemon/Model/PublicStorageFactory.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Sitecore.Support.ReleaseDaemon.Model
+{
+  using Sitecore.Diagnostics.Base;
+  using Sitecore.Support.ReleaseDaemon.Box;
+  using System.IO.Abstractions;
+
+  public class PublicStorageFactory : IPublicStorageFactory
```
Comment (AlenPelin) [inline src/ReleaseDaemon/Model/PublicStorageFactory.cs]: As far as I can see you added support of switching between several storage providers (so we can migrate from box in future). I like the idea, but it makes code harder to read. Can we hard-code single box provider instead to improve readability and supportability? i.e. completely remove `IPublicStorageFactory` and `PublicStorageFactory` and move box initialization code to `program.cs`

**Comment by olegburov** [comment]
@SC-AFI Does Box support user friendly URLs?

I would prefer to see something like this:

- https://sitecore.box.com/patches/Test.Repo.123456-8.1.3.10.zip

Instead of that:

- https://sitecore.box.com/s/0obrw5y1dmf0i42elqvg0c3jibqeovb5

**Comment by AlenPelin** [comment]
@olegburov yes, but only in this format due to [a-z0-9\-]+ requirement:
https://sitecore.box.com/v/Test-Repo-123456-8-1-3-10

**Comment by AlenPelin** [comment]
Maybe it is better to change to 
https://sitecore.box.com/v/Support-123456-813-10

@RomanChernyk @SC-AFI @alexeyromaniuha

**Comment by SC-AFI** [comment]
It looks like Box does not allow creating the Custom shared link ("vanity_url") via API.
https://docs.box.com/v2.0/reference#create-or-update-shared-link

I see 2 options:
1. Use unfriendly URL;
2. Create "Friendly" URL manually by using Box Web Application.

**Comment by AlenPelin** [comment]
I completely disagree with second option, so lets stick to first one unless we can utilize link shortifier such as bit.ly and have our links work as https://bit.ly/Sitecore-Support-123456-813-0

As far as I know, Sitecore uses it for marketing purposes so we can try to ask their API key and use their siteco.re custom domain for links like this: https://siteco.re/sitecore-support-123456-813-0 links.

**Comment by AlenPelin** [comment]
Frankly speaking, we don't really need these fancy URLs. In fact, we only need the link that looks easy to read, so we can use original ugly https://sitecore.box.com/s/0obrw5y1dmf0i42elqvg0c3jibqeovb5 URL but wrapped into an anchor tag properly like this:

This anchor [Sitecore.Support.123456-8.1.3.0](https://sitecore.box.com/s/0obrw5y1dmf0i42elqvg0c3jibqeovb5) still points to https://sitecore.box.com/s/0obrw5y1dmf0i42elqvg0c3jibqeovb5, but who cares?

**Comment by olegburov** [comment]
> It looks like Box does not allow creating the Custom shared link ("vanity_url") via API.

[Azure Blob Storage](https://azure.microsoft.com/en-us/services/storage/blobs/) supports:

- custom user friendly URLs (For example: `/patches/Sitecore-Support-123456-8.1.3-1.0`)
- custom domain (For example: `download.sitecore.net`)
- HTTPS (comes free with [Azure CDN](https://azure.microsoft.com/en-us/services/cdn/))
- CDN (with many [POP locations](https://docs.microsoft.com/en-us/azure/cdn/cdn-pop-locations) in different regions)

Is there a specific reason why we chose Box?

**Comment by AlenPelin** [comment]
@olegburov yes, corporate policy - we must use Box.

**Comment by RomanChernyk** [comment]
> This anchor Sitecore.Support.123456-8.1.3.0 still points to https://sitecore.box.com/s/0obrw5y1dmf0i42elqvg0c3jibqeovb5, but who cares?

If we are going to use the following pattern, then it would be better to provide raw url instead of a fancy anchor, as it is easier to copy.

### Permanent link
https://sitecore.box.com/s/j0vjd6zc1i1r9duj1pbgptql3o03ynlw

**Comment by AlenPelin** [comment]
Okay, let it be as is. 

Are you certain that adding `#123456-813-10` is a bad idea?

**Comment by RomanChernyk** [comment]
> Are you certain that adding #Sitecore-Support-123456-813-10 is a bad idea?

I think it may confuse customers and engineers. They may start changing the anchor in hope that it changes the patch they download.

**Comment by AlenPelin** [comment]
Do you mean change iteration number? Have you ever typed the github link manually?

**Comment by AlenPelin** [comment]
We can inform them that it is not allowed and it is for indication only.

**Comment by olegburov** [comment]
> @olegburov yes, corporate policy - we must use Box.

Asked IT folks today. They said it's a requirement for end users, but a recommendation for devs. So, if we want to use Azure Blob Storage because it provides feature that Box doesn't, they won't block us.

File: src/ReleaseDaemon/Commands/Main/030 ProcessTagCommand.cs [hunk 5]
```diff
@@ -62,7 +71,51 @@ public class ProcessTagCommand
       var release = Releases.FirstOrDefault(x => x.TagName == tagName);
       if (release != null)
       {
-        // skip tag if appropriate release already exists
+        if (release.Draft == false && release.Prerelease == false && Storage != null)
+        {
+          GetStringHash getHash = HashHelper.GetMd5Hash;
```
Comment (SC-AFI) [inline src/ReleaseDaemon/Commands/Main/030 ProcessTagCommand.cs]: I supposed that we would have necessity to replace the GetMd5Hash method with anther one. But it looks like overcomplication. So I removed the delegate and use the HashHelper.GetMd5Hash.

File: src/ReleaseDaemon/Commands/RenewPackage/030 DownloadAssetsCommand.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  {
+    public DirectoryInfoBase Dir { get; set; }
+
+    public IReleaseInfo Release { get; set; }
+
+    public OrganizationInfoBase Organization { get; set; }
+
+    public IFileSystem FileSystem { get; set; }
+
+    public async Task<IEnumerable<string>> Execute()
```
Comment (SC-AFI) [inline src/ReleaseDaemon/Commands/RenewPackage/030 DownloadAssetsCommand.cs]: I have changed the code according to the suggestion.

File: src/ReleaseDaemon/app.config (hunk 1 above)
Comment (SC-AFI) [inline src/ReleaseDaemon/app.config]: The 'app.config' file let the Box.v2.dll resolve its dependencies.

File: src/ReleaseDaemon/Model/PublicStorageInfoBase.cs (hunk 3 above)
Comment (SC-AFI) [inline src/ReleaseDaemon/Model/PublicStorageInfoBase.cs]: I suppose that we will have some general methods which should be implemented in the abstract class.

File: src/ReleaseDaemon/Model/PublicStorageFactory.cs (hunk 4 above)
Comment (SC-AFI) [inline src/ReleaseDaemon/Model/PublicStorageFactory.cs]: The 'IPublicStorageFactory' and 'PublicStorageFactory' have been removed.

File: src/ReleaseDaemon/Model/PublicStorageInfoBase.cs (hunk 3 above)
Comment (AlenPelin) [inline src/ReleaseDaemon/Model/PublicStorageInfoBase.cs]: ok
